GBA code unit testing front-end

felixjones opened the pull request:
This is a testing front end specifically for unit testing GBA code, following the suggestion from @endrift in https://github.com/mgba-emu/mgba/pull/2409#issuecomment-1005319854

At the moment I have named this frontend "mtest", but this entire PR is WIP and open for feedback and suggestions (in particular, feature suggestions).
I will say that even this basic implementation has proven remarkably lean and perfect for my needs, it has sped up my unit testing by an order of magnitude and has made writing GBA unit tests an incredibly simple task.

Basic usage is: `mgba-mtest -S 0 -R r0 my_rom.gba`

- `-S SWI` sets what SWI to exit the test on. -S 0 will exit on SWI 0. You can probably pick an unused SWI if your test uses SWI 0.
- `-R REGISTER` sets what register to read when the exit condition is met (right now, that's only -S). r0-r15, sp, lr, pc and cpsr are valid strings for this.

A basic GBA test with tonc could look like:
```c
#include <tonc.h>

static void mtest_exit(int code) {
    swi_call(0);
}

int main() {
    mtest_exit(Div(9, 3) != 3);
}
```
With a crt0/syscalls.c that calls SWI on `exit()` the code required for a GBA test is even simpler than this as the return from main would signal the test end.

Using this with CTest could look something like this:
```cmake
enable_testing()
add_test(NAME exampleTest COMMAND mgba-mtest -S 0 -R r0 exampleRom.gba)
# -S 0 uses SWI 0 as the signal
# -R r0 uses r0 as the exit code
```

And because stdout is available us usual, a ctest can regex match any of the standard output.

As of this writing (2022-06-01) the current implementation in this PR by default has no way to actually complete if the -S argument is not used. Some feedback and ideas are required here (maybe default to exiting tests after 1 frame, so test writers are forced to be specific about their GBA tests?).

An aside: libmgba is remarkably easy to work with, I am incredibly impressed with how quickly it took me to get this working during the short moments I've had today to write this code.

(PR has some garbo commits, need squashing).

endrift commented on src/platform/test/mtest-main.c:
```diff
@@ -0,0 +1,215 @@
+	"  -S SWI           Run until specified SWI call before exiting\n" \
+	"  -R REGISTER      Register to return as exit code\n" \
+
+ enum mRegisterExitCode {
+	 mRegister_NONE = -1,
+	 mRegister_GPRS = 0,
+	 mRegister_CPSR = 16
+ };
```
```suggestion
 enum mRegisterExitCode {
	 mREGISTER_NONE = -1,
	 mREGISTER_GPRS = 0,
	 mREGISTER_CPSR = 16
 };
```

to match code style, though I have to wonder why anyone would put the exit code in the CPSR. Probably not worth support.

endrift commented on src/platform/test/mtest-main.c:
```diff
@@ -0,0 +1,215 @@
+	if (core->platform(core) == mPLATFORM_GBA) {
+		((struct GBA*) core->board)->hardCrash = false;
+
+		_armSwi16 = ((struct GBA*) core->board)->cpu->irqh.swi16;
+		((struct GBA*) core->board)->cpu->irqh.swi16 = _mTestSwi16;
+		_armSwi32 = ((struct GBA*) core->board)->cpu->irqh.swi32;
+		((struct GBA*) core->board)->cpu->irqh.swi32 = _mTestSwi32;
```
Instead of hooking all SWIs and only using one, perhaps using SWI 03 (STOP) which already has a frontend-defined callback would be ideal here. See the `GBAStop` function in gba.c.

felixjones replied:
I agree that GBAStop makes a perfect choice, however my own runtime uses SWI 0 to reset when exit is called. I've left in the option to override the SWI call, but if you absolutely disagree with this then I can adjust my runtime and remove this option here

endrift commented on src/platform/test/mtest-main.c:
```diff
@@ -0,0 +1,215 @@
+
+static int _mTestRunloop(struct mCore* core, int exitSwiImmediate) {
+	do {
+		core->step(core);
```
```suggestion
		core->runLoop(core);
```
Will be much faster, and an SWI call should always break out of the runloop, especially one to GBAStop.

endrift reviewed (changes requested):
Some quick minor stuff, but in general this seems like a better approach. Coming up with an exit condition for GB ROMs is probably a good idea too. Once the overall design idea in in place I can come back with a more detailed review.

LunarLambda commented:
The devkitARM C Runtime also calls SWI 0 when main() exits, so using SWI 0 + r0 (main return value) seems really useful to support. Maybe it's overkill to support arbitrary SWI and register combinations, but 0/r0 is reasonable.

Maybe a dedicated, unused SWI would be possible? I don't know how that would interact with using an official BIOS, though. Does mGBA still hook those?

Also, the original post mentions stdout being available. How does that work? Do you mean the log interface?

felixjones commented:
> Also, the original post mentions stdout being available. How does that work? Do you mean the log interface?

Yes, mGBA logging means zero extra work needed to be done for getting a unit test to write stdout, something that CTest specifically supports
So regardless of return code, with CTest I can regex for particular stdout patterns and pass/fail/skip on match/no match.

That being said, there's probably benefit to matching memory. Probably better to actually do those matches within this test platform, rather than stdout blobs of memory for regex'ing

LunarLambda commented:
That reminds, me I wanted to write an optimized `memcmp` for libseven.

That, or having matching `.iwram.bin` and similar files that contain the expected resulting memory image and similar.

something like `test.gba` -> `test.iwram.bin` or `./expected/iwram`

endrift commented on src/platform/test/mtest-main.c:
```diff
@@ -0,0 +1,206 @@
+	_dispatchExiting = true;
+}
+
+static void _mTestSwi3Callback(struct mCore* core) {
+#ifdef M_CORE_GBA
+	_exitCode = ((struct GBA*) core->board)->cpu->regs.gprs[_returnCodeRegister];
+#endif
+	_dispatchExiting = true;
+}
```
Any reason this whole function isn't behind the ifdef?

endrift commented on src/platform/test/mtest-main.c:
```diff
@@ -0,0 +1,206 @@
+
+		if (_exitSwiImmediate == 3) {
+			// Hook into SWI 3 (shutdown)
+			callbacks = calloc(1, sizeof(*callbacks));
+			callbacks->context = core;
+			callbacks->shutdown = (void(*)(void*)) _mTestSwi3Callback;
```
Any reason this is dynamically allocated? The struct isn't exactly very big.

felixjones replied:
Probably in the mind-set of not consuming memory for GBA callbacks when running a test for GB core (which is dumb, we're hardly memory constrained).
Will remove.

endrift reviewed (approved):
Looks quite good, but I have two minor questions.

felixjones commented:
@endrift the copyright notice on the source file has my name on it from when this was a branch intended to stay private. Is there a preferred copyright holder to attribute this to?

endrift commented:
I use my (legal) name: Jeffrey Pfau. But you are more than welcome to leave your name in there.

felixjones commented:
I noticed `addCoreCallbacks` copies the `struct mCoreCallbacks`, so I moved the allocation onto the stack for that.
I updated the copyright notice for the source file.

This is in a state where I am happy with the basic functionality and I think additional features (and more importantly, GB support) could come in as their own PRs.

endrift commented on src/platform/test/mtest-main.c:
```diff
@@ -0,0 +1,202 @@
+	if (regStr[0] == 'r' || regStr[0] == 'R') {
+		++regStr;
+	}
+	return strtol(regStr, NULL, 10);
```
It would probably make sense to properly constrain this. Not that I'm too worried about some malicious actor here, but it is an obvious place to do so. Some error checking that it didn't fail to parse would probably be good too.

endrift commented on src/platform/test/CMakeLists.txt:
```diff
@@ -46,3 +46,10 @@ if(BUILD_CINEMA)
 endif()
+
+if(BUILD_MTEST)
+	add_executable(${BINARY_NAME}-mtest ${CMAKE_CURRENT_SOURCE_DIR}/mtest-main.c)
```
Given that this would wind up with a binary named mgba-mtest, I wonder if the second m is redundant. Maybe a name like mgba-test-harness, or mgba-rom-tester? It's kind of verbose, but is pretty explicit about what it is as a result.

endrift reviewed (commented):
No real objections, but a few nits and questions about how things should ideally be done.

felixjones commented:
Renamed to mgba-rom-test so it is used like a verb on the command line.
Added constraints to both the -S and -R args. -S is from 0x00 to 0xFF, -R is from R0 to R15.

endrift commented on src/platform/test/rom-test-main.c:
```diff
@@ -0,0 +1,217 @@
+	"  -S SWI           Run until specified SWI call before exiting\n" \
+	"  -R REGISTER      General purpose register to return as exit code\n" \
+
+struct romTestOpts {
```
```suggestion
struct RomTestOpts {
```

Sorry, I should have been more explicit; If m is being used as a prefix, the struct name is not initial caps, otherwise it is.

endrift commented on src/platform/test/rom-test-main.c:
```diff
@@ -0,0 +1,217 @@
+}
+
+static bool _parseSwi(const char* regStr, int* oSwi) {
+	long swi = strtol(regStr, NULL, 0);
```
It might be a good idea to make sure there isn't junk following the string, just for completeness sake. The errno alone is not sufficient given how strto* is specified.

endrift commented on src/platform/test/rom-test-main.c:
```diff
@@ -0,0 +1,217 @@
+		++regStr;
+	}
+
+	unsigned long regId = strtoul(regStr, NULL, 10);
```
Ditto.

endrift reviewed (commented):
Two small things. Sorry this is taking so long.

endrift commented on src/platform/test/rom-test-main.c:
```diff
@@ -0,0 +1,219 @@
+			// Hook into SWI 3 (shutdown)
+			struct mCoreCallbacks callbacks = {0};
+			callbacks.context = core;
+			callbacks.shutdown = (void(*)(void*)) _romTestSwi3Callback;
```
Why does this need to be cast? You can cast the void* to an mCore* inside of the callback implicitly (since it's C).

felixjones replied:
My IDE shouting at me to do explicit casts

endrift replied:
I...yes, sorry, I should have been more clear. You need the function type to match, but in the function itself you can just assign the void* argument to an mCore* variable implicitly.

endrift commented on src/platform/test/rom-test-main.c:
```diff
@@ -0,0 +1,219 @@
+		}
+	}
+#endif
+
```
It might be good to add an "unsupported" for other here, otherwise it'll just run forever (or add another way to exit, such as a maximum frame count). Not a huge deal for the initial PR though. I can add this after it's merged.

endrift reviewed (approved):
Looks good, just some simple minor nits and two questions. It will be ready to merge afterwards.

felixjones commented:
I used an `#error` macro for the no core option, my rationale being that you have to explicitly compile rom-test, so if you're going that far then you should know that you need the GBA core (for now).

endrift commented:
What I meant was that you can still compile it with GBA and GB support but if you pass a GB ROM to it you have to manually interrupt it. This isn't ideal.
